File: frontend/src/components/benchmark/content/degree_centrality/implementations/degree_centrality.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class DegreeCentrality
{
    private Dictionary<int, HashSet<int>> adjacency = new Dictionary<int, HashSet<int>>();
    private HashSet<int> nodes = new HashSet<int>();

    public void AddEdge(int u, int v)
    {
        if (!adjacency.ContainsKey(u))
            adjacency[u] = new HashSet<int>();
        if (!adjacency.ContainsKey(v))
            adjacency[v] = new HashSet<int>();

        adjacency[u].Add(v);
        adjacency[v].Add(u);
        nodes.Add(u);
        nodes.Add(v);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    public int DegreeCentrality(int node)
    {
        if (!adjacency.ContainsKey(node))
            return 0;
        return adjacency[node].Count;
    }

    public Dictionary<int, int> AllDegreeCentralities()
    {
        var result = new Dictionary<int, int>();
        foreach (int node in nodes)
        {
            result[node] = DegreeCentrality(node);
        }
        return result;
    }

    public List<(int, int)> TopKCentralNodes(int k = 10)
    {
        var centralities = AllDegreeCentralities();
        return centralities.OrderByDescending(x => x.Value)
                          .Take(k)
                          .Select(x => (x.Key, x.Value))
                          .ToList();
    }

    public double NormalizedDegreeCentrality(int node)
    {
        if (!adjacency.ContainsKey(node) || nodes.Count <= 1)
            return 0.0;
        return (double)DegreeCentrality(node) / (nodes.Count - 1);
    }

    public Dictionary<int, double> AllNormalizedDegreeCentralities()
    {
        var result = new Dictionary<int, double>();
        foreach (int node in nodes)
        {
            result[node] = NormalizedDegreeCentrality(node);
        }
        return result;
    }
}

File: frontend/src/components/benchmark/content/degree_centrality/implementations/degree_centrality.php
<?php

class DegreeCentrality {
    private array $adjacency = [];
    private array $nodes = [];

    public function addEdge(int $u, int $v): void {
        if (!isset($this->adjacency[$u])) {
            $this->adjacency[$u] = [];
        }
        if (!isset($this->adjacency[$v])) {
            $this->adjacency[$v] = [];
        }
        $this->adjacency[$u][$v] = true;
        $this->adjacency[$v][$u] = true;
        $this->nodes[$u] = true;
        $this->nodes[$v] = true;
    }

    public function buildFromEdges(array $edges): void {
        foreach ($edges as $edge) {
            $this->addEdge($edge[0], $edge[1]);
        }
    }

    public function degreeCentrality(int $node): int {
        if (!isset($this->adjacency[$node])) {
            return 0;
        }
        return count($this->adjacency[$node]);
    }

    public function allDegreeCentralities(): array {
        $result = [];
        foreach (array_keys($this->nodes) as $node) {
            $result[$node] = $this->degreeCentrality($node);
        }
        return $result;
    }

    public function topKCentralNodes(int $k = 10): array {
        $centralities = $this->allDegreeCentralities();
        arsort($centralities);
        return array_slice($centralities, 0, $k, true);
    }

    public function normalizedDegreeCentrality(int $node): float {
        if (!isset($this->adjacency[$node]) || count($this->nodes) <= 1) {
            return 0.0;
        }
        return $this->degreeCentrality($node) / (count($this->nodes) - 1);
    }

    public function allNormalizedDegreeCentralities(): array {
        $result = [];
        foreach (array_keys($this->nodes) as $node) {
            $result[$node] = $this->normalizedDegreeCentrality($node);
        }
        return $result;
    }
}

?>